### ISSUE TITLE
🎨 Color moved passages differently

### ISSUE THREAD

**Issue by nobodyinperson** [ftilmann/latexdiff#259]
`git diff --color-moved=zebra` colors *moved* lines differently so they don't show up as a huge amount of removed and added lines. Having `latexdiff` also color moved passages differently (e.g. in darkgreen) would add a lot of value to the output.

**Comment by ftilmann** [comment]
I have been thinking about this also, and miss this functionality myself but I expect it is relatively difficult to implement, particularly in the current framework and there are some subtleties involved in the definition of desired behaviour.
Do you know which algorithm is used to make the detection of moving as opposed to deleting and adding.  What happens if in addition to moving a small change is made in the moved block? Does this then get highlighted as an addition or deletion within the moved block, or does this somehow interfere with the whole block then no longer being recognised.  Is there a minimum size for the block moved?

**Comment by nobodyinperson** [comment]
> Do you know which algorithm is used to make the detection of moving as opposed to deleting and adding.

I browsed through [the `git diff` source code](https://github.com/git/git/search?q=color_moved) and quickly got lost... 😅

Python's [`difflib` also doesn't seem to have this functionality](https://bugs.python.org/issue23791), which would be a nice starting point.

> What happens if in addition to moving a small change is made in the moved block? Does this then get highlighted as an addition or deletion within the moved block, or does this somehow interfere with the whole block then no longer being recognised. Is there a minimum size for the block moved?

Good questions. The easy answer would be to „just make it configurable” and use sane defaults. 

Off the top of my head I would introduce an absolute and a relative moving thresholds. The absolute threshold would mean: „Don't consider sequences of 20 characters or less for moving”. The relative threshold would say: „If less than 10% of a moved block was also changed, still consider the whole block 'moved' and color the differences accordingly”.

**Comment by ftilmann** [comment]
Thank you for looking into this so quickly. It will be an interesting feature but I will need a block of time to think about this and implement something, and those 'blocks of time' are hard to come by these days.

**Comment by nobodyinperson** [comment]
> 'blocks of time' are hard to come by these days.

Absolutely. No pressure, I just wanted to put this idea here so it is out there.

**Comment by awillats** [comment]
Just wanted to drop by and +1 to the idea of coloring block moves differently.
It sounds like that will be non-trivial to do, so I'm not trying to add to the time pressure. But I wanted to fill in a couple of details that hopefully will make the process easier in the future.

After playing around with getting this to work on the commandline for standard difftools, indeed 
`git diff --color-moved=zebra oldfile newfile`
or 
`git diff --color-moved=plain oldfile newfile` (removes the 20 character minimum block size) both get the job done. Here's the relevant section of the [git diff docs](https://git-scm.com/docs/git-diff#Documentation/git-diff.txt---color-movedltmodegt) which answers a lot of your detail questions and could be a starting point for various design decisions.

As mentioned in [Detecting moved sections #162](https://github.com/ftilmann/latexdiff/issues/162), the core algorithm seems to be the Heckel diff algorithm described on the page for wikEd diff [Implementation](https://en.wikipedia.org/w/index.php?title=User:Cacycle/diff#Implementation). See also [Paul Heckel: A technique for isolating differences between files](https://documents.scribd.com/docs/10ro9oowpo1h81pgh1as.pdf)
Communications of the ACM 21(4):264 (1978)

Here's a couple of python implementations: [m-matelski](https://github.com/m-matelski)/[mdiff](https://github.com/m-matelski/mdiff), [lahwaacz](https://github.com/lahwaacz)/[python-wikeddiff](https://github.com/lahwaacz/python-wikeddiff), and a Stack Overflow discussion which might be helpful: [Difficulty understanding Paul Heckel's Diff Algorithm](https://stackoverflow.com/questions/42755035/difficulty-understanding-paul-heckels-diff-algorithm).

Thanks again for working on this tool, and hope these resources will be helpful at some time in the future
